install.sh

```diff
@@ -0,0 +1,408 @@
+}
+
+# Checks that the requested version is valid i.e. there is a release with the
+# name of the version supplied
+# RETURNS
+# 0  - version is valid
+# 1  - version is not valid
+validate_version() {
+  log DEBUG 'validating version'
+
+  {
+    jq -r '.[] | .name' "${releases_file}"
+    printf -- '%s' "$?" >"${temp_d}"/jq_status
+  } |
+    {
+      grep -q "${VERSION}"
+      printf -- '%s' "$?" >"${temp_d}"/grep_status
+    }
+
+  if ! grep -q '^0$' "${temp_d}"/jq_status ||
+    ! grep -q '^0$' "${temp_d}"/grep_status; then
+    log ERROR "Version ${VERSION} not found"
+    return 1
+  fi
+}
+
+# Sets the variable OS to the name of the operating system. Panics if the
+# operating system is unsupported by the CLI (i.e. missing compiled CLI
+# artifacts). Panics if not OSX or Linux compatible
+determine_operating_system() {
+  log DEBUG 'determining operating system'
+  uname_out="$(uname -a)"
+  case "${uname_out}" in
+  *Microsoft*) OS="linux" ;; #must be first since Windows subsystem for linux will have Linux in the name too
+  *microsoft*) OS="linux" ;; #WARNING: My v2 uses ubuntu 20.4 at the moment slightly different name may not always work
+  Linux*) OS="linux" ;;
+  Darwin*) OS="darwin" ;; # Apple Mac
+  *)
+    panic "Unsupported operating system. Must be Linux or Mac."
+    ;;
+  esac
+}
+
+# Sets the variable CHIPSET with the name of the current chipset architecture.
+# Panics if the chipset is not arm64 or x86
+determine_chipset() {
+  log DEBUG 'determining chipset'
+  uname_out="$(uname -m)"
+
+  case "${uname_out}" in
+  aarch64 | arm64) CHIPSET=arm64 ;;
+  x86_64*) CHIPSET=x64 ;;
+  *)
+    panic "Unsupported chipset. Supports: arm64 or x64 architectures."
+    ;;
+  esac
+}
+
+# Downloads the releases to local file
+download_releases() {
+  curl -s -L \
+    -o "${releases_file}" \
+    -H 'Accept: application/vnd.github+json' \
+    https://api.github.com/repos/companieshouse/chs-dev/releases
+}
+
+# Interrogates the release to determine the download URL for the CLI tarball
+determine_download_url() {
+  release_file="${temp_d}"/release.json
+
+  if [ "${VERSION}" = "${DEFAULT_VERSION}" ]; then
+    log DEBUG 'Fetching latest release'
+    curl -s -L -H \
+      'Accept: application/vnd.github+json' \
+      https://api.github.com/repos/companieshouse/chs-dev/releases/latest >"${release_file}"
+  else
+    log DEBUG "Looking up version ${VERSION}"
+    jq \
+      --arg version "${VERSION}" \
+      '.[] | select(.name == $version)' \
+      "${releases_file}" >"${release_file}"
+  fi
+
+  install_version="$(jq -r '.name' "${release_file}")"
+  log INFO "Installing version ${install_version}"
+
+  jq -r --arg expected_suffix "-${OS}-${CHIPSET}.tar.gz" \
+    '.assets[] | select(.name | endswith($expected_suffix)) | .browser_download_url' \
+    "${release_file}" >"${temp_d}"/download_url
+}
+
+# Downloads the tarball to temporary directory
+download_cli_tarball() {
```

Comment:
This will not, by itself, detect a failure to download. A 404 will be missed, as curl ran correctly.

Reply:
Fixed with `--force-with-body`

Reply:
There is a solution to this issue, in the current incarnation of download_file in setup_lib.bash.

Reply:
Ignore my last comment, your solution looks fine.